🐛 BUG: Cannot install @miniflare/tre@3.0.0-next.7

### What version of `Wrangler` are you using?

2.5.0

### What operating system are you using?

mac

### Describe the Bug

Several compilation errors occur during execution of ```npm install --save-dev @miniflare/tre@3.0.0-next.7```
Looks like ```better-sqlite3``` cannot be compiled.

This error wasn't present before upgrading node from 18 to 19.2. 

Full log

```
npm install --save-dev @miniflare/tre@3.0.0-next.7
npm ERR! code 1
npm ERR! path /Users/aliomarov/Desktop/uni/rjobs/node_modules/better-sqlite3
npm ERR! command failed
npm ERR! command sh -c -- prebuild-install || npm run build-release
npm ERR! > better-sqlite3@7.6.2 build-release
npm ERR! > node-gyp rebuild --release
npm ERR! 
npm ERR!   TOUCH ba23eeee118cd63e16015df367567cb043fed872.intermediate
npm ERR!   ACTION deps_sqlite3_gyp_locate_sqlite3_target_copy_builtin_sqlite3 ba23eeee118cd63e16015df367567cb043fed872.intermediate
npm ERR!   TOUCH Release/obj.target/deps/locate_sqlite3.stamp
npm ERR!   CC(target) Release/obj.target/sqlite3/gen/sqlite3/sqlite3.o
npm ERR!   LIBTOOL-STATIC Release/sqlite3.a
npm ERR!   CXX(target) Release/obj.target/better_sqlite3/src/better_sqlite3.o
npm ERR! rm ba23eeee118cd63e16015df367567cb043fed872.intermediate
npm ERR! prebuild-install warn install No prebuilt binaries found (target=19.2.0 runtime=node arch=arm64 libc= platform=darwin)
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@9.1.0
npm ERR! gyp info using node@19.2.0 | darwin | arm64
npm ERR! gyp info find Python using Python version 3.10.8 found at "/opt/homebrew/opt/python@3.10/bin/python3.10"
npm ERR! gyp info spawn /opt/homebrew/opt/python@3.10/bin/python3.10
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/aliomarov/Desktop/uni/rjobs/node_modules/better-sqlite3/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/Users/aliomarov/Library/Caches/node-gyp/19.2.0',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/Users/aliomarov/Library/Caches/node-gyp/19.2.0/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/Users/aliomarov/Desktop/uni/rjobs/node_modules/better-sqlite3',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp info spawn make
npm ERR! gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
npm ERR! In file included from ../src/better_sqlite3.cpp:4:
npm ERR! In file included from ./src/better_sqlite3.lzz:11:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/node.h:73:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8.h:24:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-array-buffer.h:12:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-local-handle.h:12:
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:476:30: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! SHARED_EXTERNAL_POINTER_TAGS(CHECK_SHARED_EXTERNAL_POINTER_TAGS)
npm ERR!                              ^
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:476:30: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:476:30: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:476:30: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:476:30: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! PER_ISOLATE_EXTERNAL_POINTER_TAGS(CHECK_NON_SHARED_EXTERNAL_POINTER_TAGS)
npm ERR!                                   ^
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:477:35: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:658:61: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR!     static_assert(kJSObjectType + 1 == kFirstJSApiObjectType);
npm ERR!                                                             ^
npm ERR!                                                             , ""
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:659:55: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR!     static_assert(kJSObjectType < kLastJSApiObjectType);
npm ERR!                                                       ^
npm ERR!                                                       , ""
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-internal.h:660:63: warning: 'static_assert' with no message is a C++17 extension [-Wc++17-extensions]
npm ERR!     static_assert(kFirstJSApiObjectType < kLastJSApiObjectType);
npm ERR!                                                               ^
npm ERR!                                                               , ""
npm ERR! In file included from ../src/better_sqlite3.cpp:4:
npm ERR! In file included from ./src/better_sqlite3.lzz:11:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/node.h:73:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8.h:24:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-array-buffer.h:13:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-object.h:9:
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-maybe.h:106:45: error: no template named 'is_lvalue_reference_v' in namespace 'std'; did you mean 'is_lvalue_reference'?
npm ERR!   template <class U, std::enable_if_t<!std::is_lvalue_reference_v<U>>*>
npm ERR!                                        ~~~~~^~~~~~~~~~~~~~~~~~~~~
npm ERR!                                             is_lvalue_reference
npm ERR! /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:910:29: note: 'is_lvalue_reference' declared here
npm ERR! struct _LIBCPP_TEMPLATE_VIS is_lvalue_reference : _BoolConstant<__is_lvalue_reference(_Tp)> { };
npm ERR!                             ^
npm ERR! In file included from ../src/better_sqlite3.cpp:4:
npm ERR! In file included from ./src/better_sqlite3.lzz:11:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/node.h:73:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8.h:24:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-array-buffer.h:13:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-object.h:9:
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-maybe.h:106:69: error: expected '(' for function-style cast or type construction
npm ERR!   template <class U, std::enable_if_t<!std::is_lvalue_reference_v<U>>*>
npm ERR!                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-maybe.h:123:43: error: no template named 'is_lvalue_reference_v' in namespace 'std'; did you mean 'is_lvalue_reference'?
npm ERR! template <class T, std::enable_if_t<!std::is_lvalue_reference_v<T>>* = nullptr>
npm ERR!                                      ~~~~~^~~~~~~~~~~~~~~~~~~~~
npm ERR!                                           is_lvalue_reference
npm ERR! /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:910:29: note: 'is_lvalue_reference' declared here
npm ERR! struct _LIBCPP_TEMPLATE_VIS is_lvalue_reference : _BoolConstant<__is_lvalue_reference(_Tp)> { };
npm ERR!                             ^
npm ERR! In file included from ../src/better_sqlite3.cpp:4:
npm ERR! In file included from ./src/better_sqlite3.lzz:11:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/node.h:73:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8.h:24:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-array-buffer.h:13:
npm ERR! In file included from /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-object.h:9:
npm ERR! /Users/aliomarov/Library/Caches/node-gyp/19.2.0/include/node/v8-maybe.h:123:67: error: expected '(' for function-style cast or type construction
npm ERR! template <class T, std::enable_if_t<!std::is_lvalue_reference_v<T>>* = nullptr>
npm ERR!                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
npm ERR! ./src/util/macros.lzz:157:21: error: no member named 'AccessorSignature' in namespace 'v8'
npm ERR!                 v8::AccessorSignature::New(isolate, recv)
npm ERR!                 ~~~~^
npm ERR! ./src/objects/database.lzz:180:21: warning: variable 'status' set but not used [-Wunused-but-set-variable]
npm ERR!                 int status = sqlite3_db_config(db_handle, SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION, 1, NULL);
npm ERR!                     ^
npm ERR! ./src/util/binder.lzz:37:51: error: no member named 'CreationContext' in 'v8::Object'
npm ERR!                 v8::Local<v8::Context> ctx = obj->CreationContext();
npm ERR!                                              ~~~~~^
npm ERR! 20 warnings and 6 errors generated.
npm ERR! make: *** [Release/obj.target/better_sqlite3/src/better_sqlite3.o] Error 1
npm ERR! gyp ERR! build error 
npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
npm ERR! gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:201:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
npm ERR! gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:293:12)
npm ERR! gyp ERR! System Darwin 21.6.0
npm ERR! gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--release"
npm ERR! gyp ERR! cwd /Users/aliomarov/Desktop/uni/rjobs/node_modules/better-sqlite3
npm ERR! gyp ERR! node -v v19.2.0
npm ERR! gyp ERR! node-gyp -v v9.1.0
npm ERR! gyp ERR! not ok
```

--- a/.changeset/grumpy-pillows-cheat.md
+++ b/.changeset/grumpy-pillows-cheat.md
@@ -1,0 +1,11 @@
+---
+"wrangler": minor
+---
+
+chore: upgrade `@miniflare/tre` to [`3.0.0-next.12`](https://github.com/cloudflare/miniflare/releases/tag/v3.0.0-next.12), incorporating changes from [`3.0.0-next.11`](https://github.com/cloudflare/miniflare/releases/tag/v3.0.0-next.11)
+
+Notably, this brings the following improvements to `wrangler dev --experimental-local`:
+
+- Adds support for Durable Objects and D1
+- Fixes an issue blocking clean exits and script reloads
+- Bumps to `better-sqlite3@8`, allowing installation on Node 19

--- a/packages/wrangler/package.json
+++ b/packages/wrangler/package.json
@@ -118,7 +118,7 @@
 		"@cloudflare/workers-types": "^4.20221111.1",
 		"@iarna/toml": "^3.0.0",
 		"@microsoft/api-extractor": "^7.28.3",
-		"@miniflare/tre": "^3.0.0-next.10",
+		"@miniflare/tre": "3.0.0-next.12",
 		"@types/better-sqlite3": "^7.6.0",
 		"@types/busboy": "^1.5.0",
 		"@types/command-exists": "^1.2.0",

--- a/packages/wrangler/src/bundle.ts
+++ b/packages/wrangler/src/bundle.ts
@@ -288,7 +288,7 @@ export async function bundleWorker(
 					currentEntry,
 					tmpDir.path,
 					betaD1Shims,
-					local,
+					local && !experimentalLocal,
 					doBindings
 				);
 			}),
@@ -816,7 +816,7 @@ async function applyD1BetaFacade(
 	entry: Entry,
 	tmpDirPath: string,
 	betaD1Shims: string[],
-	local: boolean,
+	miniflare2: boolean,
 	doBindings: DurableObjectBindings
 ): Promise<Entry> {
 	let entrypointPath = path.resolve(
@@ -867,7 +867,7 @@ async function applyD1BetaFacade(
 		],
 		define: {
 			__D1_IMPORTS__: JSON.stringify(betaD1Shims),
-			__LOCAL_MODE__: JSON.stringify(local),
+			__LOCAL_MODE__: JSON.stringify(miniflare2),
 		},
 		outfile: targetPath,
 	});

--- a/packages/wrangler/src/dev/local.tsx
+++ b/packages/wrangler/src/dev/local.tsx
@@ -231,6 +231,7 @@ function useLocalWorker({
 					enablePagesAssetsServiceBinding,
 					kvNamespaces: bindings?.kv_namespaces,
 					r2Buckets: bindings?.r2_buckets,
+					d1Databases: bindings?.d1_databases,
 					authenticatedAccountId: accountId,
 					kvRemote: experimentalLocalRemoteKv,
 					inspectorPort: runtimeInspectorPort,
@@ -776,6 +777,7 @@ export interface SetupMiniflare3Options {
 	// we need actual namespace IDs to connect to.
 	kvNamespaces: CfKvNamespace[] | undefined;
 	r2Buckets: CfR2Bucket[] | undefined;
+	d1Databases: CfD1Database[] | undefined;
 
 	// Account ID to use for authenticated Cloudflare fetch. If true, prompt
 	// user for ID if multiple available.
@@ -809,6 +811,7 @@ export async function transformMf2OptionsToMf3Options({
 	enablePagesAssetsServiceBinding,
 	kvNamespaces,
 	r2Buckets,
+	d1Databases,
 	authenticatedAccountId,
 	kvRemote,
 	inspectorPort,
@@ -835,6 +838,12 @@ export async function transformMf2OptionsToMf3Options({
 		),
 		r2Buckets: Object.fromEntries(
 			r2Buckets?.map(({ binding, bucket_name }) => [binding, bucket_name]) ?? []
+		),
+		d1Databases: Object.fromEntries(
+			d1Databases?.map(({ binding, database_id, preview_database_id }) => [
+				binding,
+				preview_database_id ?? database_id,
+			]) ?? []
 		),
 		inspectorPort,
 		verbose: true,
@@ -918,5 +927,5 @@ export async function getMiniflare3(): Promise<
 	// eslint-disable-next-line @typescript-eslint/consistent-type-imports
 	typeof import("@miniflare/tre")
 > {
-	return (miniflare3Module ??= await npxImport("@miniflare/tre@3.0.0-next.10"));
+	return (miniflare3Module ??= await npxImport("@miniflare/tre@3.0.0-next.12"));
 }

--- a/packages/wrangler/src/dev/start-server.ts
+++ b/packages/wrangler/src/dev/start-server.ts
@@ -416,6 +416,7 @@ export async function startLocalServer({
 				log,
 				kvNamespaces: bindings?.kv_namespaces,
 				r2Buckets: bindings?.r2_buckets,
+				d1Databases: bindings?.d1_databases,
 				authenticatedAccountId: accountId,
 				kvRemote: experimentalLocalRemoteKv,
 				inspectorPort,

--- a/packages/wrangler/templates/d1-beta-facade.js
+++ b/packages/wrangler/templates/d1-beta-facade.js
@@ -11,7 +11,7 @@ var D1Database = class {
 		return new D1PreparedStatement(this, query);
 	}
 	async dump() {
-		const response = await this.binding.fetch("/dump", {
+		const response = await this.binding.fetch("http://d1/dump", {
 			method: "POST",
 			headers: {
 				"content-type": "application/json",
@@ -79,7 +79,7 @@ var D1Database = class {
 						params,
 				  }
 		);
-		const response = await this.binding.fetch(endpoint, {
+		const response = await this.binding.fetch(new URL(endpoint, "http://d1"), {
 			method: "POST",
 			headers: {
 				"content-type": "application/json",

--- a/packages/wrangler/templates/middleware/middleware-miniflare3-json-error.ts
+++ b/packages/wrangler/templates/middleware/middleware-miniflare3-json-error.ts
@@ -1,15 +1,27 @@
 import type { Middleware } from "./common";
 
+interface JsonError {
+	message?: string;
+	name?: string;
+	stack?: string;
+	cause?: JsonError;
+}
+
+function reduceError(e: any): JsonError {
+	return {
+		name: e?.name,
+		message: e?.message ?? String(e),
+		stack: e?.stack,
+		cause: e?.cause === undefined ? undefined : reduceError(e.cause),
+	};
+}
+
 // See comment in `bundle.ts` for details on why this is needed
 const jsonError: Middleware = async (request, env, _ctx, middlewareCtx) => {
 	try {
 		return await middlewareCtx.next(request, env);
 	} catch (e: any) {
-		const error = {
-			name: e?.name,
-			message: e?.message ?? String(e),
-			stack: e?.stack,
-		};
+		const error = reduceError(e);
 		return Response.json(error, {
 			status: 500,
 			headers: { "MF-Experimental-Error-Stack": "true" },

